ktor-server/ktor-server-plugins/ktor-server-sessions/jvm/src/io/ktor/server/sessions/KotlinxSessionSerializer.kt

```diff
@@ -0,0 +1,397 @@
+    }
+
+    override fun decodeDouble(): Double {
+        return parameters[currentName]!!.drop(2).toDouble()
+    }
+
+    override fun decodeFloat(): Float {
+        return parameters[currentName]!!.drop(2).toFloat()
+    }
+
+    override fun decodeInt(): Int {
+        return parameters[currentName]!!.drop(2).toInt()
+    }
+
+    override fun decodeLong(): Long {
+        return parameters[currentName]!!.drop(2).toLong()
+    }
+
+    override fun decodeString(): String {
+        return parameters[currentName]!!.drop(2)
+    }
+
+    override fun decodeNotNullMark(): Boolean {
+        return parameters[currentName]!! != "#n"
+    }
+
+    override fun decodeNull(): Nothing? {
+        return null
+    }
+
+    override fun decodeEnum(enumDescriptor: SerialDescriptor): Int {
+        val enumName = decodeString()
+        val index = enumDescriptor.getElementIndex(enumName)
+        if (index == CompositeDecoder.UNKNOWN_NAME) {
+            throw IllegalStateException(
+                "${enumDescriptor.serialName} does not contain element with name '$enumName'"
+            )
+        }
+        return index
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+private class ListLikeDecoder(
+    override val serializersModule: SerializersModule,
+    string: String,
+) : AbstractDecoder() {
+
+    private var currentIndex = -1
+    private val items = string.split("&")
+
+    override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
+        if (++currentIndex == items.size) {
+            return CompositeDecoder.DECODE_DONE
+        }
+        return currentIndex
+    }
+
+    override fun decodeBoolean(): Boolean {
+        return items[currentIndex] == "#bot"
+    }
+
+    override fun decodeChar(): Char {
+        return items[currentIndex][3]
+    }
+
+    override fun decodeDouble(): Double {
+        return items[currentIndex].drop(2).toDouble()
+    }
+
+    override fun decodeFloat(): Float {
+        return items[currentIndex].drop(2).toFloat()
+    }
+
+    override fun decodeInt(): Int {
+        return items[currentIndex].drop(2).toInt()
+    }
+
+    override fun decodeLong(): Long {
+        return items[currentIndex].drop(2).toLong()
+    }
+
+    override fun decodeString(): String {
+        return items[currentIndex].drop(2)
+    }
+
+    override fun decodeNotNullMark(): Boolean {
+        return items[currentIndex] != "#n"
+    }
+
+    override fun decodeNull(): Nothing? {
+        return null
+    }
+
+    override fun decodeEnum(enumDescriptor: SerialDescriptor): Int {
+        val enumName = decodeString()
+        val index = enumDescriptor.getElementIndex(enumName)
+        if (index == CompositeDecoder.UNKNOWN_NAME) {
+            throw IllegalStateException(
+                "${enumDescriptor.serialName} does not contain element with name '$enumName'"
+            )
+        }
+        return index
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class, InternalSerializationApi::class)
```

Comment:
why `InternalSerializationApi` is used here?
Not sure it's a good idea. If it's not possible, IMO it would be better to reduce scope of this annotation per declaration, instead of class

Reply:
`NamedValueDecoder` is marked as `InternalSerializationApi`. It simplifies a lot parsing data into a map. Another option would be to replicate it in our codebase, which I would like to avoid. `MapDecoder` is private, so it's relatively safe to have it here.

Reply:
the main issue is if `NamedValueDecoder` will be changed in not backward compatible maner (like adding some function without default implementation, which is safe for kx.serialization, as it's internal API) it can cause backward compatibility issues for ktor.

Reply:
This class is relatively stable and it should be ok to rely on it here in the private API.